fix(StatusQ): Dynamic width calculation for StatusMenu

alexjba opened the pull request:
### What does the PR do
Closing: #11028 

Calculating the `StatusMenu` implicitWidth based on content width.
<!-- Fill in the relevant information below to help us evaluate your proposed changes. -->

### Affected areas
`StatusMenu`
<!-- List the affected areas (e.g wallet, browser, etc..) -->

### Screenshot of functionality (including design for comparison)

<img width="577" alt="Screenshot 2023-06-15 at 13 35 36" src="https://github.com/status-im/status-desktop/assets/47811206/b08fcfed-5202-462e-a606-c6ace0d6b2e9">

Checking for regressions:
<img width="287" alt="Screenshot 2023-06-15 at 13 37 03" src="https://github.com/status-im/status-desktop/assets/47811206/fe560dc9-2cb4-48c7-bc09-506d28e238f6">
<img width="287" alt="Screenshot 2023-06-15 at 13 37 28" src="https://github.com/status-im/status-desktop/assets/47811206/73ab1cc2-2eec-4c7e-957d-a4ead7fed6c9">

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 5f0651cf | [#1](https://ci.status.im/job/status-desktop/job/branches/job/linux/job/x86_64/job/tests-imports/job/PR-11106/1/) | 2023-06-15 10:45:03 | ~5 min | `tests/imports` | [:page_facing_up:`log`](https://ci.status.im/job/status-desktop/job/branches/job/linux/job/x86_64/job/tests-imports/job/PR-11106/1//consoleText) |

caybro commented on ui/StatusQ/src/StatusQ/Popups/StatusMenuItem.qml:
```diff
@@ -10,7 +10,6 @@ import StatusQ.Popups 0.1
 MenuItem {
     id: root
 
-    implicitWidth: parent ? parent.width : 0
```
aka "too much code"

caybro replied:
Actually, on a second thought, we might want to set some absolute max value, so that the menu width doesn't explode with overly long strings

alexjba replied:
I was thinking on that and didn't add a max value because I think elided menu items are just bugs, so the right fix IMHO would be to change the phrasing to something appropriate for menu items.

But I'm happy to do it if you think it's better to have it.

caybro replied:
Yeah that makes sense except for when the content gets generated, eg a wallet address?

alexjba replied:
I guess a default of 640 would do? I've updated the PR.

caybro reviewed (approved):
Nice :rocket:

igor-sirotin commented on ui/StatusQ/src/StatusQ/Popups/StatusMenu.qml:
```diff
@@ -88,6 +88,11 @@ Menu {
         visible: root.hideDisabledItems ? enabled : true
         height: visible ? implicitHeight : 0
+        onImplicitWidthChanged: {
+            var adjustedImplicitWidth = implicitWidth + 2 * padding
```
```suggestion
            const adjustedImplicitWidth = implicitWidth + leftPadding + rightPadding
```

micieslak replied:
There is no need to add `StatusMenuItem`'s padding here because it's already included in the `implicitWidth` of that component.
So it can be just:
```
root.contentWidth = Math.max(root.contentWidth, implicitWidth)
```
![Screenshot from 2023-06-15 15-51-05](https://github.com/status-im/status-desktop/assets/20650004/c2dc4bfe-6ac3-47d4-9651-2418cb273d65)

igor-sirotin commented on ui/StatusQ/src/StatusQ/Popups/StatusMenu.qml:
```diff
@@ -88,6 +88,11 @@ Menu {
         visible: root.hideDisabledItems ? enabled : true
         height: visible ? implicitHeight : 0
+        onImplicitWidthChanged: {
+            var adjustedImplicitWidth = implicitWidth + 2 * padding
```
I'm not sure, but is this part even needed?

`Menu` is a `Popup`. And it's `contentWidth` should be calculated automatically based on `contentItem` implicit width (which is a `StatusListView` in our case).

And I suppose that `StatusListView` should calculate it's contentWidth from `implicitWidth` of all items 🤔 

So the real change is the one you did in `StatusMenuItem`, but it wasn't enough?

igor-sirotin commented on ui/StatusQ/src/StatusQ/Popups/StatusMenu.qml:
```diff
@@ -88,6 +88,11 @@ Menu {
     delegate: StatusMenuItem {
         visible: root.hideDisabledItems ? enabled : true
         height: visible ? implicitHeight : 0
```
btw, probably worth adding `width: ListView.view.width` here?

micieslak replied:
> btw, probably worth adding `width: ListView.view.width` here?

Probably not bc it creates some kind of circular dependency. `ListView.view.width` depends on `Menu`'s `contentWidth`, which depends on `StatusMenuItem` implicit width (max along all instances). Even if it's not harmful it's not needed imo.

micieslak reviewed (changes requested):
Nice solution! Idea seems to work fine, but there are some details to fix. First one is in the comment regarding paddings.

The second is regarding various `MenuItems` in the app with `width` set to specific value. It won't work well with transactions and exact values should be removed from all places.

I found that problem in menus in several files:
```
ProfileContextMenu
UserStatusContextMenu
ProfileDialogView
EditCroppedImagePanel
ProfileHeader
```
But please take a look once again and test, I could overlook sth.

alexjba commented:
> Nice solution! Idea seems to work fine, but there are some details to fix. First one is in the comment regarding paddings.
> 
> The second is regarding various `MenuItems` in the app with `width` set to specific value. It won't work well with transactions and exact values should be removed from all places.
> 
> I found that problem in menus in several files:
> 
> ```
> ProfileContextMenu
> UserStatusContextMenu
> ProfileDialogView
> EditCroppedImagePanel
> ProfileHeader
> ```
> 
> But please take a look once again and test, I could overlook sth.

Will do, thanks! I've just seen it won't work well with horizontal layouts either (while the default implementation is fine). IDK if it makes sense to have horizontal StatusMenu tho 🤔.

alexjba commented:
> I found that problem in menus in several files:
```
ProfileContextMenu
UserStatusContextMenu
ProfileDialogView
EditCroppedImagePanel
ProfileHeader
```

Fixed. I've removed some of the static values, but others are there for UX reasons. We also use StatusMenu as a drop-down menu. In this case it needs to have the same width as the parent.


> And I suppose that StatusListView should calculate it's contentWidth from implicitWidth of all items 🤔

Fixed. The ListView will use the implicit width calculated as max delegate implicit width. I've used helper properties to decouple `delegate` and `contentItem` implementations because any of them could be replaced.

alexjba commented:
> LGTM 👍

@igor-sirotin Sorry, I've just re-asked for review before getting the approve update.

micieslak commented on ui/StatusQ/src/StatusQ/Popups/StatusMenu.qml:
```diff
@@ -85,15 +88,28 @@ Menu {
         currentIndex: root.currentIndex
-        implicitHeight: contentHeight
-        implicitWidth: contentWidth
+        implicitHeight: orientation === ListView.Horizontal ? d.maxDelegateImplHeight : contentHeight
```
the `orientation` of the list is not exposed directly from the `StatusMenu` so imo it's not necessary to assume that sb can change that (that list view is internal impl detail). So that check can be removed or orientation exposed as an alias

micieslak reviewed (approved):
Nice!
